### PR TITLE
Initial implementation of an JSON array based Set

### PR DESCRIPTION
**Pull request by simonbasle**
@daschl this is ready for your review :)

### CODE REVIEW

File: src/main/java/com/couchbase/client/commons/CouchbaseArraySet.java [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+                data.add(o);
+            }
+        }
+        bucket.upsert(JsonArrayDocument.create(id, data));
+    }
+
+    @Override
+    public int size() {
+        //TODO use subdoc GET_COUNT when available
+        JsonArrayDocument current = bucket.get(id, JsonArrayDocument.class);
```
Comment (daschl) [inline src/main/java/com/couchbase/client/commons/CouchbaseArraySet.java:84]: I wonder if stuff like this and empty() all of those that are similar could be refactored into a helper class? so once GET_COUNT and friends come along we only fix it in one place?

File: src/main/java/com/couchbase/client/commons/CouchbaseArraySet.java [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+
+    @Override
+    public boolean add(T t) {
+        if (t == null) {
+            throw new NullPointerException("Unsupported null value");
+        }
+
+        for (int i = 0; i < MAX_OPTIMISTIC_LOCKING_ATTEMPTS; i++) {
+            try {
+                JsonArrayDocument current = bucket.get(id, JsonArrayDocument.class);
```
Comment (daschl) [inline src/main/java/com/couchbase/client/commons/CouchbaseArraySet.java:119]: this is similar to contains, could we refactor that into a helper method private?

File: src/main/java/com/couchbase/client/commons/CouchbaseArraySet.java [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+            } catch (CASMismatchException e) {
+                //retry
+            }
+        }
+        throw new ConcurrentModificationException("Couldn't perform remove in less than " + MAX_OPTIMISTIC_LOCKING_ATTEMPTS + " iterations");
+    }
+
+    @Override
+    public void clear() {
+        bucket.upsert(JsonArrayDocument.create(id, JsonArray.empty()));
```
Comment (daschl) [inline src/main/java/com/couchbase/client/commons/CouchbaseArraySet.java:183]: I think this is also duplicated right?

File: src/main/java/com/couchbase/client/commons/CouchbaseArraySet.java (hunk 2 above)
Comment (simonbasle) [reply src/main/java/com/couchbase/client/commons/CouchbaseArraySet.java:119]: it is not exactly like contains as we have to extract the cas and take it into account for the subsequent append. That said, maybe I could use ARRAY_ADD_UNIQUE... but then it has an even greater limitation (cannot accept anything that is non-primitive, like `JsonObject` or `JsonArray`).

**Comment by daschl** [comment]
👍

**Pull request merged**